diff --git a/include/vsync/atomic/internal/arm64.h b/include/vsync/atomic/internal/arm64.h
index 7e92c35..339b47f 100644
--- a/include/vsync/atomic/internal/arm64.h
+++ b/include/vsync/atomic/internal/arm64.h
@@ -364,6 +364,7 @@ vatomic32_await_eq(const vatomic32_t *a, vuint32_t v)
         ".align 5\n"
         "1:\n"
         "wfe\n"
+        "prfm pldl1krrp, [%[a]]\n"
         "ldaxr %w[val], %[a]\n"
         "cmp %w[val],  %w[exp]\n"
         "b.ne 1b\n"
